File: sdks/python/src/mcparmor/_pool.py
"""ArmoredPool — pre-spawned pool of ArmoredProcess instances."""

import asyncio
import logging
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from pathlib import Path
from types import TracebackType

from mcparmor._process import ArmoredProcess

_logger = logging.getLogger(__name__)


class ArmoredPoolError(OSError):
    """Raised when a pool operation fails."""


class ArmoredPool:
    """
    Manages a pool of warm :class:`ArmoredProcess` instances.

    Pre-spawns *size* processes at :meth:`start` time and hands them out
    via :meth:`acquire` / :meth:`release`. Designed for workloads that
    maintain many concurrent tool processes (e.g. Arqitect's 50-process pool).

    Usage as an async context manager::

        async with ArmoredPool(
            command=["python", "tool.py"],
            armor="./armor.json",
            size=10,
        ) as pool:
            proc = await pool.acquire()
            try:
                result = proc.invoke({"method": "run", "params": {}})
            finally:
                await pool.release(proc)

    Args:
        command: The tool command and arguments for each process.
        armor: Path to the armor.json manifest.
        size: Number of processes to pre-spawn.
        ready_signal: If True, each process waits for a ``{"ready": true}``
            line from stdout before becoming available.
        profile: Profile name override.
        no_os_sandbox: Disable OS-level sandbox enforcement.
        cwd: Working directory for spawned processes.
    """

    def __init__(
        self,
        command: list[str],
        *,
        armor: str | Path | None = None,
        size: int = 4,
        ready_signal: bool = False,
        profile: str | None = None,
        no_os_sandbox: bool = False,
        cwd: str | Path | None = None,
    ) -> None:
        if size < 1:
            raise ValueError("Pool size must be at least 1.")

        self._command = command
        self._armor = armor
        self._size = size
        self._ready_signal = ready_signal
        self._profile = profile
        self._no_os_sandbox = no_os_sandbox
        self._cwd = cwd

        self._available: asyncio.Queue[ArmoredProcess] = asyncio.Queue(maxsize=size)
        self._all: list[ArmoredProcess] = []
        self._lock = asyncio.Lock()
        self._started = False
        self._closed = False

    # ------------------------------------------------------------------
    # Async context manager
    # ------------------------------------------------------------------

    async def __aenter__(self) -> "ArmoredPool":
        """Start the pool and return it."""
        await self.start()
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        """Close all pooled processes."""
        await self.close()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def start(self) -> None:
        """
        Pre-spawn all processes in the pool.

        Each process is started as a persistent :class:`ArmoredProcess`
        (context-manager style). If *ready_signal* was set, each process
        waits for the ready handshake before joining the available queue.

        Raises:
            ArmoredPoolError: If the pool has already been started or closed.
        """
        async with self._lock:
            if self._closed:
                raise ArmoredPoolError("Pool has been closed.")
            if self._started:
                raise ArmoredPoolError("Pool is already started.")

            for _ in range(self._size):
                proc = self._create_process()
                proc._persistent = True
                proc._spawn()
                if self._ready_signal:
                    proc.wait_ready()
                self._all.append(proc)
                self._available.put_nowait(proc)

            self._started = True
            _logger.info("ArmoredPool started with %d processes", self._size)

    async def acquire(self) -> ArmoredProcess:
        """
        Acquire an available process from the pool.

        Blocks until a process becomes available via :meth:`release`.

        Returns:
            A running :class:`ArmoredProcess` ready for :meth:`~ArmoredProcess.invoke`.

        Raises:
            ArmoredPoolError: If the pool is not started or has been closed.
        """
        if not self._started:
            raise ArmoredPoolError("Pool is not started. Call start() first.")
        if self._closed:
            raise ArmoredPoolError("Pool has been closed.")
        return await self._available.get()

    async def release(self, proc: ArmoredProcess) -> None:
        """
        Return a process to the pool.

        If the process has died (``is_alive()`` is False), it is replaced
        with a freshly spawned one.

        Args:
            proc: The process previously obtained via :meth:`acquire`.

        Raises:
            ArmoredPoolError: If the pool has been closed.
        """
        if self._closed:
            proc.close()
            return

        if not proc.is_alive():
            _logger.warning("Replacing dead process (pid=%s) in pool", proc.pid)
            proc.close()
            replacement = self._create_process()
            replacement._persistent = True
            replacement._spawn()
            if self._ready_signal:
                replacement.wait_ready()

            async with self._lock:
                self._all = [p if p is not proc else replacement for p in self._all]
            proc = replacement

        await self._available.put(proc)

    async def close(self) -> None:
        """
        Close all processes in the pool.

        Safe to call multiple times; subsequent calls are no-ops.
        """
        async with self._lock:
            if self._closed:
                return
            self._closed = True

        for proc in self._all:
            proc.close()
        self._all.clear()
        _logger.info("ArmoredPool closed")

    @property
    def size(self) -> int:
        """Return the configured pool size."""
        return self._size

    @property
    def available(self) -> int:
        """Return the number of currently available processes."""
        return self._available.qsize()

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _create_process(self) -> ArmoredProcess:
        """Create a new ArmoredProcess with this pool's configuration."""
        return ArmoredProcess(
            command=self._command,
            armor=self._armor,
            profile=self._profile,
            no_os_sandbox=self._no_os_sandbox,
            cwd=self._cwd,
        )

File: sdks/node/src/pool.ts
/**
 * ArmoredPool — pre-spawned pool of ArmoredProcess instances.
 *
 * Manages a fixed-size collection of warm processes for workloads that
 * maintain many concurrent tool subprocesses (e.g. Arqitect's 50-process pool).
 */

import { ArmoredProcess, ArmoredProcessError, type ArmoredProcessOptions } from './process.ts';

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

/** Thrown on pool-level errors (start, acquire, release, close). */
export class ArmoredPoolError extends Error {
  constructor(message: string, options?: { cause?: unknown }) {
    super(message, options);
    this.name = 'ArmoredPoolError';
  }
}

// ---------------------------------------------------------------------------
// Pool options
// ---------------------------------------------------------------------------

/** Configuration for {@link ArmoredPool}. */
export interface ArmoredPoolOptions {
  /** The tool command and arguments for each process. */
  readonly command: readonly string[];
  /** Path to the armor.json capability manifest. */
  readonly armor?: string;
  /** Number of processes to pre-spawn. Default: 4. */
  readonly size?: number;
  /**
   * If true, each process waits for a `{"ready": true}` JSON line
   * from stdout before becoming available.
   */
  readonly readySignal?: boolean;
  /** Override the base profile declared in armor.json. */
  readonly profile?: string;
  /** Disable OS-level sandbox (skips Layer 2 enforcement). */
  readonly noOsSandbox?: boolean;
  /** Working directory for spawned processes. */
  readonly cwd?: string;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_POOL_SIZE = 4;
const MIN_POOL_SIZE = 1;

// ---------------------------------------------------------------------------
// ArmoredPool class
// ---------------------------------------------------------------------------

/**
 * A fixed-size pool of warm {@link ArmoredProcess} instances.
 *
 * Pre-spawns *size* processes at {@link start} time and hands them out via
 * {@link acquire} / {@link release}.
 *
 * @example
 * ```typescript
 * const pool = new ArmoredPool({
 *   command: ['node', 'tool.js'],
 *   armor: './armor.json',
 *   size: 10,
 * });
 * await pool.start();
 *
 * const proc = await pool.acquire();
 * try {
 *   const result = await proc.invoke({ method: 'run', params: {} });
 * } finally {
 *   pool.release(proc);
 * }
 *
 * await pool.close();
 * ```
 */
export class ArmoredPool {
  readonly #options: ArmoredPoolOptions;
  readonly #size: number;
  readonly #available: ArmoredProcess[] = [];
  readonly #all: ArmoredProcess[] = [];
  readonly #waiters: Array<(proc: ArmoredProcess) => void> = [];
  #started = false;
  #closed = false;

  /**
   * Create a new ArmoredPool.
   *
   * @param options - Pool configuration.
   * @throws {RangeError} If size is less than 1.
   */
  constructor(options: ArmoredPoolOptions) {
    const size = options.size ?? DEFAULT_POOL_SIZE;
    if (size < MIN_POOL_SIZE) {
      throw new RangeError('Pool size must be at least 1');
    }
    this.#options = options;
    this.#size = size;
  }

  // ------------------------------------------------------------------
  // Public API
  // ------------------------------------------------------------------

  /**
   * Pre-spawn all processes in the pool.
   *
   * If `readySignal` was set, each process waits for the ready handshake
   * before joining the available queue.
   *
   * @throws {ArmoredPoolError} If the pool is already started or closed.
   */
  async start(): Promise<void> {
    if (this.#started) {
      throw new ArmoredPoolError('Pool is already started');
    }
    if (this.#closed) {
      throw new ArmoredPoolError('Pool has been closed');
    }

    const spawnPromises: Array<Promise<ArmoredProcess>> = [];
    for (let i = 0; i < this.#size; i++) {
      spawnPromises.push(this.#createProcess());
    }

    const processes = await Promise.all(spawnPromises);
    for (const proc of processes) {
      this.#all.push(proc);
      this.#available.push(proc);
    }

    this.#started = true;
  }

  /**
   * Acquire an available process from the pool.
   *
   * If no process is currently available, the returned promise blocks
   * until one is released via {@link release}.
   *
   * @returns A running {@link ArmoredProcess}.
   * @throws {ArmoredPoolError} If the pool is not started or has been closed.
   */
  async acquire(): Promise<ArmoredProcess> {
    if (!this.#started) {
      throw new ArmoredPoolError('Pool is not started. Call start() first');
    }
    if (this.#closed) {
      throw new ArmoredPoolError('Pool has been closed');
    }

    const proc = this.#available.shift();
    if (proc !== undefined) {
      return proc;
    }

    return new Promise<ArmoredProcess>((resolve) => {
      this.#waiters.push(resolve);
    });
  }

  /**
   * Return a process to the pool.
   *
   * If the process has died, it is replaced with a freshly spawned one.
   *
   * @param proc - The process previously obtained via {@link acquire}.
   */
  release(proc: ArmoredProcess): void {
    if (this.#closed) {
      void proc.close();
      return;
    }

    if (!proc.isAlive()) {
      void proc.close();
      void this.#replaceProcess(proc);
      return;
    }

    this.#handOff(proc);
  }

  /**
   * Close all processes in the pool.
   *
   * Safe to call multiple times; subsequent calls are no-ops.
   */
  async close(): Promise<void> {
    if (this.#closed) {
      return;
    }
    this.#closed = true;

    const closePromises = this.#all.map((proc) => proc.close());
    await Promise.all(closePromises);
    this.#all.length = 0;
    this.#available.length = 0;
  }

  /** The configured pool size. */
  get size(): number {
    return this.#size;
  }

  /** The number of currently available processes. */
  get available(): number {
    return this.#available.length;
  }

  // ------------------------------------------------------------------
  // Private helpers
  // ------------------------------------------------------------------

  /**
   * Create and spawn a single ArmoredProcess.
   *
   * @returns A running process, optionally ready-signalled.
   */
  async #createProcess(): Promise<ArmoredProcess> {
    const processOptions: ArmoredProcessOptions = {
      command: this.#options.command,
      ...(this.#options.armor !== undefined ? { armor: this.#options.armor } : {}),
      ...(this.#options.profile !== undefined ? { profile: this.#options.profile } : {}),
      ...(this.#options.noOsSandbox !== undefined ? { noOsSandbox: this.#options.noOsSandbox } : {}),
      ...(this.#options.cwd !== undefined ? { cwd: this.#options.cwd } : {}),
      ...(this.#options.readySignal !== undefined ? { readySignal: this.#options.readySignal } : {}),
    };
    return ArmoredProcess.spawn(processOptions);
  }

  /**
   * Replace a dead process in the pool with a fresh one.
   *
   * @param dead - The process that died.
   */
  async #replaceProcess(dead: ArmoredProcess): Promise<void> {
    const replacement = await this.#createProcess();
    const index = this.#all.indexOf(dead);
    if (index !== -1) {
      this.#all[index] = replacement;
    } else {
      this.#all.push(replacement);
    }
    this.#handOff(replacement);
  }

  /**
   * Hand a process to a waiting acquirer, or put it in the available queue.
   *
   * @param proc - The process to hand off.
   */
  #handOff(proc: ArmoredProcess): void {
    const waiter = this.#waiters.shift();
    if (waiter !== undefined) {
      waiter(proc);
    } else {
      this.#available.push(proc);
    }
  }
}
